pkg/util/worker_test.go

```diff
@@ -0,0 +1,171 @@
+package util
+
+import (
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"go.uber.org/atomic"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+)
+
+func newTestAsyncWorker(reconcileFunc ReconcileFunc) *asyncWorker {
+	options := Options{
+		Name:          "test_async_worker",
+		KeyFunc:       MetaNamespaceKeyFunc,
+		ReconcileFunc: reconcileFunc,
+		RateLimiterOptions: ratelimiterflag.Options{
+			RateLimiterBaseDelay:  time.Millisecond,
+			RateLimiterMaxDelay:   10 * time.Millisecond,
+			RateLimiterQPS:        5000,
+			RateLimiterBucketSize: 100,
+		},
+	}
+	worker := NewAsyncWorker(options)
+
+	return worker.(*asyncWorker)
+}
+
+func Test_asyncWorker_Enqueue(t *testing.T) {
+	const name = "fake_node"
+
+	worker := newTestAsyncWorker(nil)
+
+	node := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+
+	worker.Enqueue(node)
+
+	item, _ := worker.queue.Get()
+
+	if name != item {
+		t.Errorf("Added Item: %v, want: %v", item, name)
+	}
+}
+
+func Test_asyncWorker_AddAfter(t *testing.T) {
+	const name = "fake_node"
+	const duration = 1 * time.Second
+
+	worker := newTestAsyncWorker(nil)
+
+	start := time.Now()
+	worker.AddAfter(name, duration)
+
+	item, _ := worker.queue.Get()
+	end := time.Now()
+
+	if name != item {
+		t.Errorf("Added Item: %v, want: %v", item, name)
+	}
+
+	elapsed := end.Sub(start)
+	if elapsed < duration {
+		t.Errorf("Added Item should dequeued after %v, but the actually elapsed time is %v.",
+			duration.String(), elapsed.String())
+	}
+}
+
+type asyncWorkerReconciler struct {
+	lock sync.Mutex
+	// guarded by lock
+	shouldPass map[int]struct{}
+}
+
+func newAsyncWorkerReconciler() *asyncWorkerReconciler {
+	return &asyncWorkerReconciler{
+		shouldPass: make(map[int]struct{}),
+	}
+}
+
+func (a *asyncWorkerReconciler) ReconcileFunc(key QueueKey) error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	v := key.(int)
+	if _, ok := a.shouldPass[v]; !ok {
+		// every item retry once
+		a.shouldPass[v] = struct{}{}
+		return errors.New("fail once")
+	}
+
+	return nil
+}
+
+func (a *asyncWorkerReconciler) ProcessedItem() map[int]struct{} {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	ret := make(map[int]struct{}, len(a.shouldPass))
+	for k, v := range a.shouldPass {
+		ret[k] = v
+	}
+
+	return ret
+}
+
+func Test_asyncWorker_Run(t *testing.T) {
+	const cnt = 2000
+
+	reconcile := newAsyncWorkerReconciler()
+	worker := newTestAsyncWorker(reconcile.ReconcileFunc)
+	stopChan := make(chan struct{})
+	worker.Run(5, stopChan)
+
+	for i := 0; i < cnt; i++ {
+		worker.Add(i)
+	}
+
+	time.Sleep(time.Second)
```

Comment:
We can wait for `len(processed) == cnt` for a period of time, we don't know if 1s is enough or not, you know the test will be launched on various environments.

Reply:
I already set ratelimiter QPS & delay, so ratelimiter will not delay the execution. I think 1s is enough for most such short code to finish.

Reply:
@RainbowMango

Reply:
I think I can write a polling logic: check the status periodically, fail after certain seconds. For example, check every 50 ms, fail after 10s. This way, bigger wait duration won't result in bigger test execution time.

Reply:
> I think I can write a polling logic: check the status periodically, fail after certain seconds. For example, check every 50 ms, fail after 10s. This way, bigger wait duration won't result in bigger test execution time.

Done